packages/react-scripts/scripts/eject.js

```diff
@@ -224,7 +233,7 @@ inquirer
       }
     }
 
-    if (fs.existsSync(paths.yarnLockFile)) {
+    if (fs.existsSync(paths.yarnLockFile) || isYarnAvailable()) {
```

Comment:
If the problem is due to npm corrupting a Yarn-produced tree, this can be solved by deleting `node_modules` first before re-running npm.

Reply:
i don't think it's corrupting, per-se -- it's just that this is happening within a yarn workspace, so npm can't handle it properly.

Reply:
But I’d like to understand what’s going on here better. Why do we run Yarn in this test even if the user created an app with npm?

Reply:
You mean ```echo yes | npm run eject```, right?  We can change that to yarn.  In that case, maybe you want to detect that the eject script was run with yarn and automatically use yarn for the install?  That would help enforce some consistency with yarn/npm usage, but not sure what your philosophy is on that.  Let me know what you want to do in this PR, not sure this PR is the appropriate place for that auto-detecting change.

Reply:
Hmm no, I just mean that if an app was created with npm, we should run npm on ejecting, and if it was created with Yarn, we should use Yarn. 

We use the lockfile to tell if the app was created with Yarn. Is this not sufficient in this case? Why?

Reply:
Apparently yarn only creates a lockfile at the top-level of the workspace, see https://github.com/yarnpkg/yarn/issues/4206.

Reply:
That makes sense. But it also means our check is inadequate (?) Seems like if we’re inside a workspace we also want to run Yarn. Can we detect this?

Reply:
Yeah, you're right.  Here's how we currently detect: https://github.com/facebook/create-react-app/blob/47d2d94118db8d8366ac8036469439885d3e1525/packages/react-scripts/config/paths.js#L139-L143

Reply:
Let's extract that into `react-dev-utils` maybe and use in both places?
Or cache this in `config/paths.js`.

Reply:
There are a bunch of places where a decision to use yarn or npm is made, probably all should be updated:
https://github.com/facebook/create-react-app/blob/c34267786f33498123dfdb872f51f33bbf7a1bd4/packages/create-react-app/createReactApp.js#L188
https://github.com/facebook/create-react-app/blob/c34267786f33498123dfdb872f51f33bbf7a1bd4/packages/react-scripts/scripts/init.js#L88
https://github.com/facebook/create-react-app/blob/c34267786f33498123dfdb872f51f33bbf7a1bd4/packages/react-scripts/scripts/eject.js#L227
https://github.com/facebook/create-react-app/blob/c34267786f33498123dfdb872f51f33bbf7a1bd4/packages/react-scripts/scripts/build.js#L48

Reply:
The first one intentionally checks for existence of Yarn because we haven't created a project yet.

All the other ones are checking for presence of `yarn.lock`. If it's committed, then the project was *created* with Yarn (maybe by another member of your team) so you should really install it (even if you don't have it).

Reply:
For createReactApp.js: if a user is creating an app in a yarn workspace, they really want to be using yarn, using npm will cause issues.

Reply:
This sounds good.